### PR TITLE
Onboarding ledger missing states

### PR DESCRIPTION
**Pull request by arjunu**

### CODE REVIEW

File: packages/web/app/components/shared/WarningAlert.module.scss [hunk 1]
```diff
@@ -7,13 +7,13 @@
   justify-content: center;
   text-align: center;
   color: $neu-flamingo-red;
+  margin-bottom: px-to-rem(45px);
```
Comment (desfero) [inline packages/web/app/components/shared/WarningAlert.module.scss:10]: can you check the visual regression, please? Warning still looks broken

Comment (arjunu) [reply packages/web/app/components/shared/WarningAlert.module.scss:10]: fixed

File: packages/web/intl/locales/en-en.json [hunk 2]
```diff
@@ -87,6 +87,7 @@
   "checkUnverifiedEmailModal.button-text": "Okay",
   "checkUnverifiedEmailModal.light-wallet": "We've sent a welcome email containing an email verification link to {unverifiedEmail}. To continue with the account setup, please confirm your email address.",
   "checkUnverifiedEmailModal.title": "Please check your inbox",
+  "clear-search": "Clear search",
```
Comment (desfero) [inline packages/web/intl/locales/en-en.json:90]: hm, let's stick to the convetion? `common.clear-search` or do a ledger flow specific string

File: packages/web/webpack/webpack.config.dev.js [hunk 3]
```diff
@@ -27,6 +27,7 @@ module.exports = merge.smart(configCommon, {
   mode: "development",
   devtool: "cheap-module-source-map",
   devServer: {
+    clientLogLevel: "warn",
```
Comment (desfero) [inline packages/web/webpack/webpack.config.dev.js:30]: +1

File: packages/shared-utils/src/utils/Validators.spec.ts [hunk 4]
```diff
@@ -2,21 +2,30 @@ import { expect } from "chai";
...
       "44'/62'/0'/0",
       "44'/60'/0'/0a",
       "a44'/60'/0'/0",
       "44'/60'/0'/0'/s",
+      "44'/62'/0'/0'",
     ];
 
     it("should return null for correct prefix path", () => {
-      expect(derivationPathPrefixValidator(correctPrefix)).to.be.null;
+      for (const item of correctPrefixes) {
```
[the remark this answers is not in the document]
Comment (arjunu) [reply packages/shared-utils/src/utils/Validators.spec.ts:26]: updated

File: packages/shared-utils/src/utils/Validators.ts [hunk 5]
```diff
@@ -1,8 +1,8 @@
-const DERIVATION_PATH_PREFIX_REGEX = /^(44'\/6[0|1]'\/\d+'?\/(?:[0|1]\/)?)$/;
+const DERIVATION_PATH_PREFIX_REGEX = /^(44'\/6[0|1]'\/\d+'?\/(\d+)?)$/;
 
 export const derivationPathPrefixValidator = (dp: string): string | null => {
   if (!DERIVATION_PATH_PREFIX_REGEX.test(dp)) {
-    return "Invalid derivation path prefix! Example of valid ones are \"44'/60'/0'\" (ledger) or \"44'/60'/0'/0\" (BIP44)";
+    return `Please search using a supported derivation path prefix such as 44'/60'/0 (Ledger) or 44'/60'/0'/0 (BIP44)`;
```
[the remark this answers is not in the document]
Comment (arjunu) [reply packages/shared-utils/src/utils/Validators.ts:5]: updated to return boolean result

File: packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/TableControls.tsx [hunk 6]
```diff
@@ -1,92 +1,100 @@
...
+    const errorMessage = derivationPathPrefixValidator(
+      DEFAULT_DERIVATION_PATH_SUB_PREFIX_1 + derivationPathPrefix,
+    );
     this.props.setErrorMessage(errorMessage);
 
-    if (errorMessage) {
-      this.props.onDerivationPathPrefixError();
-    } else {
-      this.props.onDerivationPathPrefixChange(fixedDp);
+    if (!errorMessage) {
```
[the remark this answers is not in the document]
Comment (arjunu) [reply packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/TableControls.tsx:43]: code changed

File: packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/TableControls.tsx [hunk 7]
```diff
@@ -1,92 +1,100 @@
...
+            className={cn(styles.input, { [styles.error]: !!errorMessage })}
             name="derivationPathPrefix"
             value={derivationPathPrefix}
-            onChange={this.onDerivationPathPrefixChange}
-            placeholder={formatIntlMessage(
-              "wallet-selector.ledger.derivation-path-selector.placeholder",
-            )}
+            onChange={event => setDerivationPathPrefix(event.target.value)}
+            onKeyDown={event => {
+              event.keyCode === 13 && this.onSearch();
```
[the remark this answers is not in the document]
Comment (arjunu) [reply packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/TableControls.tsx:77]: changed to form

File: packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/TableControls.tsx [hunk 8]
```diff
@@ -1,92 +1,100 @@
...
         </div>
 
+        <Button
+          layout={EButtonLayout.PRIMARY}
+          size={EButtonSize.NORMAL}
+          width={EButtonWidth.NO_PADDING}
+          className={styles.searchButton}
+          onClick={this.onSearch}
+        >
+          <InlineIcon svgIcon={magnifyingGlass} className={styles.searchIcon} />
```
[the remark this answers is not in the document]
Comment (arjunu) [reply packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/TableControls.tsx:89]: added

File: packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/TableControls.tsx [hunk 9]
```diff
@@ -1,92 +1,100 @@
...
+        </Button>
+
+        {derivationPathPrefix && derivationPathPrefix !== DEFAULT_DERIVATION_PATH_SUB_PREFIX_2 && (
```
[the remark this answers is not in the document]
Comment (arjunu) [reply packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/TableControls.tsx:92]: moved

File: packages/web/intl/locales/en-en.json (hunk 2 above)
Comment (arjunu) [reply packages/web/intl/locales/en-en.json:90]: changed

File: packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/TableControls.tsx [hunk 10]
```diff
@@ -1,92 +1,100 @@
...
-  onDerivationPathPrefixError: () => void;
+  onSearch: (derivationPathPrefix: string) => void;
   setDerivationPathPrefix: (derivationPathPrefix: string) => void;
   setErrorMessage: (errorMessage: string | null) => void;
   showPrevAddresses: () => void;
   showNextAddresses: () => void;
   hasPreviousAddress: boolean;
   derivationPathPrefix: string;
-  errorMessage: string | null;
+  errorMessage: React.ReactNode;
```
[the remark this answers is not in the document]
Comment (arjunu) [reply packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/TableControls.tsx:31]: changed

File: packages/web/app/modules/wallet-selector/ledger-wizard/sagas.ts [hunk 11]
```diff
@@ -176,10 +176,11 @@ export function* loadLedgerAccountsEffect({
   yield put(actions.walletSelector.setLedgerAccounts(accounts, derivationPathPrefix));
 }
 
-export function* loadLedgerAccounts(_: TGlobalDependencies): Generator<any, any, any> {
+export function* loadLedgerAccounts({ logger }: TGlobalDependencies): Generator<any, any, any> {
   try {
     yield neuCall(loadLedgerAccountsEffect);
   } catch (e) {
+    logger.error(e);
```
[the remark this answers is not in the document]
Comment (arjunu) [reply packages/web/app/modules/wallet-selector/ledger-wizard/sagas.ts:183]: added

File: packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/WalletLedgerChooserBase.tsx [hunk 12]
```diff
@@ -1,71 +1,113 @@
...
   let content;
 
-  if (errorMessage) {
+  if (!advanced) {
+    if (loading) {
+      content = <LoadingIndicator />;
+    } else {
+      content = (
+        <PanelRounded>
+          <div>
```
[the remark this answers is not in the document]
Comment (arjunu) [reply packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/WalletLedgerChooserBase.tsx:77]: updated

File: packages/web/app/components/shared/WarningAlert.module.scss (hunk 1 above)
Comment (arjunu) [reply packages/web/app/components/shared/WarningAlert.module.scss:10]: removed

File: packages/web/app/components/shared/WarningAlert.module.scss [hunk 13]
```diff
@@ -7,13 +7,14 @@
...
 
   .icon {
     @include rect(20px);
     display: block;
-    margin-right: 5px;
-    margin-left: auto;
-    margin-bottom: px-to-rem(45px);
+    margin-right: px-to-rem(5px);
+    align-self: flex-start;
```
[the remark this answers is not in the document]
Comment (arjunu) [reply packages/web/app/components/shared/WarningAlert.module.scss:16]: center aligned

File: packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/TableControls.module.scss [hunk 14]
```diff
@@ -58,3 +62,24 @@
   color: $neu-light-blue-grey-two;
   cursor: pointer;
 }
+
+.search-button {
+  height: 40px;
```
[the remark this answers is not in the document]
Comment (arjunu) [reply packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/TableControls.module.scss:67]: removed height, how can I avoid `min-width`?

File: packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/TableControls.module.scss [hunk 15]
```diff
@@ -5,20 +5,33 @@
...
   width: px-to-rem(320px);
-  padding-left: px-to-rem(40px);
-  padding-right: px-to-rem(40px);
+  padding-left: px-to-rem(63px);
+  padding-right: px-to-rem(20px);
+  background-color: transparent;
+
+  &.error {
+    color: $neu-flamingo-red;
+    border: 2px solid $neu-flamingo-red;
```
[the remark this answers is not in the document]
Comment (arjunu) [reply packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/TableControls.module.scss:26]: postponed

File: packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/TableControls.module.scss [hunk 16]
```diff
@@ -58,3 +62,24 @@
...
 }
+
+.search-button {
+  height: 40px;
+  min-width: 40px;
+  margin-right: px-to-rem(12px);
+}
+
+.search-icon {
+  width: px-to-rem(24px);
```
[the remark this answers is not in the document]
Comment (arjunu) [reply packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/TableControls.module.scss:73]: postponed

File: packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/TableControls.module.scss (hunk 14 above)
[the remark this answers is not in the document]
Comment (arjunu) [reply packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/TableControls.module.scss:67]: changed

File: packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/TableControls.tsx [hunk 17]
```diff
@@ -1,92 +1,100 @@
...
       showNavigation,
       setDerivationPathPrefix,
+      errorMessage,
     } = this.props;
 
     return (
       <div className={styles.wrapper}>
         <div className={styles.inputWrapper}>
-          <InlineIcon svgIcon={magnifyingGlass} className={styles.searchIcon} />
+          <span className={styles.prefix}>{DEFAULT_DERIVATION_PATH_SUB_PREFIX_1}</span>
```
[the remark this answers is not in the document]
Comment (arjunu) [reply packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/TableControls.tsx:68]: ok

File: packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/TableControls.tsx [hunk 18]
```diff
@@ -1,92 +1,100 @@
...
+          size={EButtonSize.NORMAL}
+          width={EButtonWidth.NO_PADDING}
+          className={styles.searchButton}
+          onClick={this.onSearch}
+        >
+          <InlineIcon svgIcon={magnifyingGlass} className={styles.searchIcon} />
+        </Button>
+
+        {derivationPathPrefix && derivationPathPrefix !== DEFAULT_DERIVATION_PATH_SUB_PREFIX_2 && (
+          <span className={styles.clearSearch} onClick={this.onClearSearch}>
```
[the remark this answers is not in the document]
Comment (arjunu) [reply packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/TableControls.tsx:93]: ButtonLink is used for links, added tabIndex here

File: packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/WalletLedgerChooserBase.tsx [hunk 19]
```diff
@@ -1,71 +1,113 @@
...
   showPrevAddresses,
   showNextAddresses,
   loading,
   advanced,
-  onDerivationPathPrefixChange,
-  onDerivationPathPrefixError,
-  handleAdvanced,
+  onSearch,
+  toggleAdvanced,
+  searchError,
```
[the remark this answers is not in the document]
Comment (arjunu) [reply packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/WalletLedgerChooserBase.tsx:48]: changed

File: packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/WalletLedgerChooserBase.tsx [hunk 20]
```diff
@@ -1,71 +1,113 @@
...
 }) => {
   const [derivationPathPrefix, setDerivationPathPrefix] = React.useState<string>(
-    DEFAULT_DERIVATION_PATH_PREFIX,
+    DEFAULT_DERIVATION_PATH_SUB_PREFIX_2,
   );
+
   const [errorMessage, setErrorMessage] = React.useState<string | null>(null);
+  let error;
```
[the remark this answers is not in the document]
Comment (arjunu) [reply packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/WalletLedgerChooserBase.tsx:55]: it's simplified now

File: packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/WalletLedgerChooserBase.tsx [hunk 21]
```diff
@@ -1,71 +1,113 @@
...
+  const onBackToDefaults = () => {
+    toggleAdvanced();
+    setDerivationPathPrefix(DEFAULT_DERIVATION_PATH_SUB_PREFIX_2);
+    setErrorMessage(null);
+  };
 
   let content;
 
-  if (errorMessage) {
+  if (!advanced) {
```
[the remark this answers is not in the document]
Comment (arjunu) [reply packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/WalletLedgerChooserBase.tsx:71]: done

File: packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/TableControls.module.scss [hunk 22]
```diff
@@ -58,3 +62,24 @@
...
+  height: px-to-rem(24px);
+  color: transparent;
+}
+
+.clear-search {
+  color: $neu-shuttle-gray;
+  border-bottom: 1px solid $neu-gull-gray;
```
[the remark this answers is not in the document]
Comment (arjunu) [reply packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/TableControls.module.scss:80]: didn't move

File: packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/WalletLedgerChooserBase.module.scss [hunk 23]
```diff
@@ -36,6 +36,13 @@
   font-weight: 600;
 }
 
-.error-message {
-  padding: px-to-rem(20px);
+div {
+  &.error-message {
```
[the remark this answers is not in the document]
Comment (arjunu) [reply packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/WalletLedgerChooserBase.module.scss:40]: removed

File: packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/TableControls.tsx (hunk 18 above)
[the remark this answers is not in the document]
Comment (desfero) [reply packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/TableControls.tsx]: ok let's keep it as it is. But in general tabindex is not enough as you still need to handle manually key press for enter and space (see `useButtonRole.ts`). We can see whether Button with `EButtonLayout.LINK` looks the same

File: packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/TableControls.tsx [hunk 24]
```diff
@@ -1,92 +1,100 @@
...
   setErrorMessage: (errorMessage: string | null) => void;
   showPrevAddresses: () => void;
   showNextAddresses: () => void;
   hasPreviousAddress: boolean;
   derivationPathPrefix: string;
-  errorMessage: string | null;
+  errorMessage: React.ReactNode;
 }
 
 class TableControlsBase extends React.Component<ITableControlsBaseProps & IIntlProps, {}> {
```
[the remark this answers is not in the document]
Comment (arjunu) [reply packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/TableControls.tsx]: removed

File: packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/TableControls.tsx (hunk 18 above)
[the remark this answers is not in the document]
Comment (arjunu) [reply packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/TableControls.tsx]: EButtonLayout.LINK looks very different, I ended up using the hook

File: packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/TableControls.module.scss [hunk 25]
```diff
@@ -58,3 +62,24 @@
...
+  cursor: pointer;
+  font-family: $font-family-roboto;
+  font-size: px-to-rem(12);
```
[the remark this answers is not in the document]
Comment (arjunu) [reply packages/web/app/components/wallet-selector/shared/ledger-wallet/WalletLedgerChooser/TableControls.module.scss:82]: used

**Pull request merged**